test/run_test.py

```diff
@@ -424,21 +430,57 @@ def calculate_shards(num_shards: int, tests: List[str], job_times: Dict[str, Tup
     return sharded_jobs
 
 
-def get_shard(which_shard: int, num_shards: int, tests: List[str]) -> List[str]:
+def pull_job_times_from_S3() -> Dict[str, Tuple[float, int]]:
     if HAVE_BOTO3:
-        s3_reports = get_test_time_reports_from_S3()
+        try:
+            s3_reports = get_test_time_reports_from_S3()
+        except (RuntimeError, RuntimeWarning):
+            print('Ran into issues downloading reports from S3.')
+            s3_reports = []
     else:
-        print('Please install boto3 to enable automatic test sharding.')
+        print('Please install boto3 to enable using S3 test times for automatic sharding and test categorization.')
         s3_reports = []
+
     if len(s3_reports) == 0:
-        print('Gathered no reports from S3. Proceeding with default sharding plan.')
+        print('Gathered no reports from S3. Please proceed without them.')
+        return dict()
+
+    return calculate_job_times(s3_reports)
+
+
+def get_shard(which_shard: int, num_shards: int, tests: List[str]) -> List[str]:
+    jobs_to_times = pull_job_times_from_S3()
+
+    # Got no stats from S3, returning early to save runtime
+    if len(jobs_to_times) == 0:
+        print('Gathered no stats from S3. Proceeding with default sharding plan.')
         return tests[which_shard - 1 :: num_shards]
-    jobs_to_times = calculate_job_times(s3_reports)
+
     shards = calculate_shards(num_shards, tests, jobs_to_times)
     _, tests_from_shard = shards[which_shard - 1]
     return tests_from_shard
 
 
+def get_slow_tests_based_on_S3() -> List[str]:
+    jobs_to_times = pull_job_times_from_S3()
```

Comment:
This is the same exact call as line 452. For simplicity reasons and to not introduce a new global variable, I allowed this as the same code would only be run twice if --determine-from and --shard are both specified.

However, I acknowledge that the two options are run together quite frequently, and so I am down to either:
- store its results in another global variable that can be called from both methods OR
- compute its results in the main function and pass it through arguments to both methods OR 
- another smarter solution I haven't thought of yet

Which would be preferred?

Reply:
what would be the downside(s) of solution 2?

Reply:
It's not as readable--the jobs_to_times dictionary would be passed down through get_selected_tests(), which would only need it for the sharding part.